### ISSUE TITLE
Elastic-Agent is not using specified --proxy-url and stays indefinitely in Updating status

### ISSUE THREAD

**Issue by morashwan** [elastic/elastic-agent#2447]
**Issue Summary** 

- In 8.6.2 and 8.7.0 --proxy-url is not persisted after installation. Agent checks in and remains indefinitely in Updating status (see below for more details). As the agent cannot download the policy, the diagnostics is not really helpful, e.g. computed-config.yaml is empty. The Agent can still check in though and diagnostics bundle can be created through the UI. For me it looks like the agent uses the proxy to check in but not download the policy.
- Tested on CentOS and Windows, same behaviour
- Internal Slack Discussion [Link](https://elastic.slack.com/archives/C01QQ449KE1/p1680186912753469?thread_ts=1678383819.454419&cid=C01QQ449KE1)
- This seems to be related to https://github.com/elastic/elastic-agent/issues/2304

**Elastic Agent installation with Proxy (outbound https blocked)**
```
sudo ./elastic-agent install --url=https://<redacted>.fleet.europe-west4.gcp.elastic-cloud.com:443 --enrollment-token=<redacted> --proxy-url http://<redacted>:3128
Elastic Agent will be installed at /opt/Elastic/Agent and will run as a service. Do you want to continue? [Y/n]:Y
{"log.level":"info","@timestamp":"2023-04-04T09:58:56.522Z","log.origin":{"file.name":"cmd/enroll_cmd.go","file.line":475},"message":"Starting enrollment to URL: https://655e492f271d4fd9b6c55b6b2fcc9921.fleet.europe-west4.gcp.elastic-cloud.com:443/","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-04-04T09:58:57.850Z","log.origin":{"file.name":"cmd/enroll_cmd.go","file.line":273},"message":"Successfully triggered restart on running Elastic Agent.","ecs.version":"1.6.0"}
Successfully enrolled the Elastic Agent.
```

**Elastic Agent Status after installation (stays indefinitely in this status)**

```
# ./elastic-agent status
State: STARTING
Message: Waiting for initial configuration and composable variables
Fleet State: HEALTHY
Fleet Message: (no message)
Components: (none)
```

**Agent Status in Fleet** - Agent can still check in at regular intervals, [diagnostics bundle](https://github.com/elastic/elastic-agent/files/11147272/elastic-agent-diagnostics-2023-04-04T10-07-01Z-00.zip) can be downloaded but  Agent remains in Updating indefinitely
  
<img width="1274" alt="image" src="https://user-images.githubusercontent.com/11572702/229758970-c711c8c9-5fb1-476d-b7df-8f40de73a237.png">

**Debugging this behaviour further**, I found out elastic-agent is using two ports for communication https 443 (connection cannot be established as 443 is blocked) & proxy 3128 (connection established through the proxy)

```
sudo netstat -nputw | grep 'agent' | grep -v 127.0.0.1
tcp        0      1 10.164.0.114:60562      34.91.16.60:443         SYN_SENT    25725/elastic-agent
tcp        0      0 10.164.0.114:53554      34.147.34.141:3128      ESTABLISHED 25725/elastic-agent
tcp        0      1 10.164.0.114:60614      34.91.16.60:443         SYN_SENT    25725/elastic-agent
tcp        0      1 10.164.0.114:60578      34.91.16.60:443         SYN_SENT    25725/elastic-agent
tcp        0      1 10.164.0.114:60536      34.91.16.60:443         SYN_SENT    25725/elastic-agent
tcp        0      1 10.164.0.114:60588      34.91.16.60:443         SYN_SENT    25725/elastic-agent
tcp        0      1 10.164.0.114:60544      34.91.16.60:443         SYN_SENT    25725/elastic-agent
tcp        0      1 10.164.0.114:60622      34.91.16.60:443         SYN_SENT    25725/elastic-agent
tcp        0      1 10.164.0.114:60596      34.91.16.60:443         SYN_SENT    25725/elastic-agent
```

**Stopping the proxy -> Agent goes offline**
Just to confirm the Agent is using the Proxy/ checking in through the Proxy, I stopped the proxy. As expected the agent goes then offline after timing out and the errors appear in the log `proxyconnect tcp: dial tcp redacted:3128: connect: connection refused`

<img width="1221" alt="image" src="https://user-images.githubusercontent.com/11572702/229774426-bff70d7c-3ef4-49a5-bdf1-937156568807.png">

```
{"log.level":"warn","@timestamp":"2023-04-04T11:03:57.113Z","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":193},"message":"Possible transient error during checkin with fleet-server, retrying","log":{"source":"elastic-agent"},"error":{"message":"fail to checkin to fleet-server: all hosts failed: 1 error occurred:\n\t* requester 0/1 to host https://redacted.fleet.europe-west4.gcp.elastic-cloud.com:443/ errored: Post \"https://redacted.fleet.europe-west4.gcp.elastic-cloud.com:443/api/fleet/agents/d9ef22eb-5f49-4a68-b2f2-45d3c479ae15/checkin?\": proxyconnect tcp: dial tcp redacted:3128: connect: connection refused\n\n"},"request_duration_ns":4799920,"failed_checkins":2,"retry_after_ns":210512771622,"ecs.version":"1.6.0"}
```

**Proxy setting in the Fleet Proxy UI**
Setting the Proxy in the Fleet Proxy UI and assigning to Fleet (without reinstallation) brings the Agent back to Healthy and Policy can be downloaded. This seems for me also as a confirmation that the proxy is used but not consistently used. This would not solve the issue here unfortunately as it is not possible to assign different proxies for the same Fleet Server (Agents in different networks communicating with the same Fleet Server). Also port 443 disappears which is expected if the proxy is used

```
sudo netstat -nputw | grep 'agent' | grep -v 127.0.0.1
tcp        0      0 10.164.0.114:58128      34.147.34.141:3128      ESTABLISHED 25725/elastic-agent
tcp        0      0 10.164.0.114:57528      34.147.34.141:3128      ESTABLISHED 25725/elastic-agent
tcp        0      0 10.164.0.114:58134      34.147.34.141:3128      ESTABLISHED 25725/elastic-agent
```

**Comment by pierrehilbert** [comment]
We are working on a fix.

**Comment by cmacknz** [comment]
Since this affects 8.6.2 as well this is likely caused by the same bug as https://github.com/elastic/elastic-agent/issues/2304

@AndersonQ once you fix https://github.com/elastic/elastic-agent/issues/2304 can you confirm that this issue is not reproducible either?

What I suspect is happening is that on the first checkin the persisted proxy configuration from `--proxy-url` is being overwritten with the empty one received from Fleet. This would be consistent with the fact that reverting https://github.com/elastic/elastic-agent/pull/2172 fixes https://github.com/elastic/elastic-agent/issues/2304.

**Comment by AndersonQ** [comment]
> What I suspect is happening is that on the first checkin the persisted proxy configuration from `--proxy-url` is being overwritten with the empty one received from Fleet. This would be consistent with the fact that reverting #2172 fixes #2304.

Yes, this seems to be the behaviour, and it might be the expected behaviour. We'll discuss this behaviour today withe the team.

So far the recommendation is to have the proxy in the policy first, then install/enrol the agents

**Comment by cmacknz** [comment]
The result of the discussion is captured in https://github.com/elastic/elastic-agent/issues/2304#issuecomment-1497944166

**Comment by Danouchka** [comment]
Hi @pierrehilbert , @AndersonQ   Is the issue solved in Elastic-Agent 8.6.2 ? I have a customer who needs deploy 8.6.2-x elastic-agents. For now, he's stuck in 8.5.3

**Comment by AndersonQ** [comment]
Hi, yes, the fix has been merged and will be released in 8.7.1

**Comment by Danouchka** [comment]
Hi @AndersonQ I said that we would need the fix on 8.6.2 as well as my customer wont move for now to 8.7

**Comment by AndersonQ** [comment]
No, 8.6.2 was released long ago. The fix is on 8.8 and backported to 8.7.1
